README.md
# LLM powered development for VSCode

**llm-vscode** is an extension for all things LLM. It uses [**llm-ls**](https://github.com/huggingface/llm-ls) as its backend.

We also have extensions for:
* [neovim](https://github.com/huggingface/llm.nvim)
* [jupyter](https://github.com/bigcode-project/jupytercoder)

Previously **huggingface-vscode**.

## Features

### Code completion

This plugin supports "ghost-text" code completion, à la Copilot.

### Choose your model

Requests for code generation are made via an HTTP request.

You can use the Hugging Face [Inference API](https://huggingface.co/inference-api) or your own HTTP endpoint, provided it adheres to the API specified [here](https://huggingface.co/docs/api-inference/detailed_parameters#text-generation-task) or [here](https://huggingface.github.io/text-generation-inference/#/Text%20Generation%20Inference/generate).

The list of officially supported models is located in the config template section.

### Always fit within the context window

The prompt sent to the model will always be sized to fit within the context window, with the number of tokens determined using [tokenizers](https://github.com/huggingface/tokenizers).

### Code attribution

Hit `Cmd+shift+a` to check if the generated code is in in [The Stack](https://huggingface.co/datasets/bigcode/the-stack).
This is a rapid first-pass attribution check using [stack.dataportraits.org](https://stack.dataportraits.org).
We check for sequences of at least 50 characters that match a Bloom filter.
This means false positives are possible and long enough surrounding context is necesssary (see the [paper](https://dataportraits.org/) for details on n-gram striding and sequence length).
[The dedicated Stack search tool](https://hf.co/spaces/bigcode/search) is a full dataset index and can be used for a complete second pass. 

## Installation

Install like any other [vscode extension](https://marketplace.visualstudio.com/items?itemName=HuggingFace.huggingface-vscode).

By default, this extension uses [bigcode/starcoder](https://huggingface.co/bigcode/starcoder) & [Hugging Face Inference API](https://huggingface.co/inference-api) for the inference.

#### HF API token

You can supply your HF API token ([hf.co/settings/token](https://hf.co/settings/token)) with this command:
1. `Cmd/Ctrl+Shift+P` to open VSCode command palette
2. Type: `Llm: Login`

If you previously logged in with `huggingface-cli login` on your system the extension will read the token from disk.

## Configuration

You can check the full list of configuration settings by opening your settings page (`cmd+,`) and typing `Llm`.

### Endpoint

You can configure the endpoint to which requests will be sent.

Let's say your current code is this:
```py
import numpy as np
import scipy as sp
{YOUR_CURSOR_POSITION}
def hello_world():
    print("Hello world")
```

The request body will then look like:
```js
const inputs = `{start token}import numpy as np\nimport scipy as sp\n{end token}def hello_world():\n    print("Hello world"){middle token}`
const data = { inputs, parameters: { max_new_tokens: 256 } };

const model = configuration.modelIdOrEndpoint;
let endpoint;
if (model.startswith("https://")) {
  endpoint = model;
} else {
  endpoint = `https://api-inference.huggingface.co/models/${model}`;
}

const res = await fetch(endpoint, {
    body: JSON.stringify(data),
    headers,
    method: "POST"
});

const json = await res.json() as { generated_text: string };
```

Note that the example above is a simplified version to explain what is happening under the hood.

### [**llm-ls**](https://github.com/huggingface/llm-ls)

By default, **llm-ls** is bundled with the extension. When developing locally or if you built your own binary because your platform is not supported, you can set the `llm.lsp.binaryPath` setting to the path of the binary.

### Tokenizer

**llm-ls** uses [**tokenizers**](https://github.com/huggingface/tokenizers) to make sure the prompt fits the `context_window`.

To configure it, you have a few options:
* No tokenization, **llm-ls** will count the number of characters instead:
```json
{
  "llm.tokenizer": null
}
```
* from a local file on your disk:
```json
{
  "llm.tokenizer": {
    "path": "/path/to/my/tokenizer.json"
  }
}
```
* from a Hugging Face repository, **llm-ls** will attempt to download `tokenizer.json` at the root of the repository:
```json
{
  "llm.tokenizer": {
    "repository": "myusername/myrepo"
  }
}
```
* from an HTTP endpoint, **llm-ls** will attempt to download a file via an HTTP GET request:
```json
{
  "llm.tokenizer": {
    "url": "https://my-endpoint.example.com/mytokenizer.json",
    "to": "/download/path/of/mytokenizer.json"
  }
}
```

### Code Llama

To test Code Llama 13B model:
1. Make sure you have the [latest version of this extension](#installing).
2. Make sure you have [supplied HF API token](#hf-api-token)
3. Open Vscode Settings (`cmd+,`) & type: `Llm: Config Template`
4. From the dropdown menu, choose `codellama/CodeLlama-13b-hf`

Read more [here](https://huggingface.co/blog/codellama) about Code LLama.

### Phind and WizardCoder

To test [Phind/Phind-CodeLlama-34B-v2](https://hf.co/Phind/Phind-CodeLlama-34B-v2) and/or [WizardLM/WizardCoder-Python-34B-V1.0](https://hf.co/WizardLM/WizardCoder-Python-34B-V1.0) :
1. Make sure you have the [latest version of this extension](#installing).
2. Make sure you have [supplied HF API token](#hf-api-token)
3. Open Vscode Settings (`cmd+,`) & type: `Llm: Config Template`
4. From the dropdown menu, choose `Phind/Phind-CodeLlama-34B-v2` or `WizardLM/WizardCoder-Python-34B-V1.0`

Read more about Phind-CodeLlama-34B-v2 [here](https://huggingface.co/Phind/Phind-CodeLlama-34B-v2) and WizardCoder-15B-V1.0 [here](https://huggingface.co/WizardLM/WizardCoder-15B-V1.0).

## Developing

1. Clone this repo: `git clone https://github.com/huggingface/llm-vscode`
2. Install deps: `cd llm-vscode && npm i`
3. In vscode, open `Run and Debug` side bar & click `Launch Extension`

## Community

| Repository | Description |
| --- | --- |
| [huggingface-vscode-endpoint-server](https://github.com/LucienShui/huggingface-vscode-endpoint-server) | Custom code generation endpoint for this repository |

CHANGELOG.md
# Change Log

All notable changes to the "llm-vscode" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [Unreleased]

- Initial release

test.py
def main();
    

src/runCompletion.ts
import { Position, Range, TextDocument, WorkspaceConfiguration, workspace, window, env, Uri } from "vscode";
import {URL} from "url";
import fetch from "node-fetch";
import type { Config as HFCodeConfig } from "./configTemplates"
import { PREFIX, SUFFIX } from "./configTemplates"
import { AutocompleteResult, ResultEntry } from "./binary/requests/requests";
import { CHAR_LIMIT, FULL_BRAND_REPRESENTATION } from "./globals/consts";
import languages from "./globals/languages";
import { setDefaultStatus, setLoadingStatus } from "./statusBar/statusBar";
import { logInput, logOutput } from "./outputChannels";
import { getTabnineExtensionContext } from "./globals/tabnineExtensionContext";

export type CompletionType = "normal" | "snippet";

let didShowTokenWarning = false;

export default async function runCompletion(
  document: TextDocument,
  position: Position,
  timeout?: number,
  currentSuggestionText = ""
): Promise<AutocompleteResult | null | undefined> {
  setLoadingStatus(FULL_BRAND_REPRESENTATION);
  const offset = document.offsetAt(position);
  const beforeStartOffset = Math.max(0, offset - CHAR_LIMIT);
  const afterEndOffset = offset + CHAR_LIMIT;
  const beforeStart = document.positionAt(beforeStartOffset);
  const afterEnd = document.positionAt(afterEndOffset);
  const prefix =  document.getText(new Range(beforeStart, position)) + currentSuggestionText;
  const suffix = document.getText(new Range(position, afterEnd));

  const config = workspace.getConfiguration("HuggingFaceCode") as WorkspaceConfiguration & HFCodeConfig;
  const { modelIdOrEndpoint, isFillMode, autoregressiveModeTemplate, fillModeTemplate, stopTokens, tokensToClear, temperature, maxNewTokens } = config;

  const context = getTabnineExtensionContext();
  const apiToken = await context?.secrets.get("apiToken");

  let endpoint = ""
  try{
    new URL(modelIdOrEndpoint);
    endpoint = modelIdOrEndpoint;
  }catch(e){
    endpoint = `https://api-inference.huggingface.co/models/${modelIdOrEndpoint}`

    // if user hasn't supplied API Token yet, ask user to supply one
    if(!apiToken && !didShowTokenWarning){
      didShowTokenWarning = true;
      void window.showInformationMessage(`In order to use "${modelIdOrEndpoint}" through Hugging Face API Inference, you'd need Hugging Face API Token`,
        "Get your token"
      ).then(clicked => {
        if (clicked) {
          void env.openExternal(Uri.parse("https://github.com/huggingface/huggingface-vscode#hf-api-token"));
        }
      });
    }
  }

  // use FIM (fill-in-middle) mode if suffix is available
  const inputs = (isFillMode && suffix.trim()) ? fillModeTemplate.replace(PREFIX, prefix).replace(SUFFIX, suffix) : autoregressiveModeTemplate.replace(PREFIX, prefix).replace(SUFFIX, suffix);

  const data = {
    inputs,
    parameters: {
      max_new_tokens: clipMaxNewTokens(maxNewTokens as number),
      temperature,
      do_sample: temperature > 0,
      top_p: 0.95,
      stop: stopTokens
    }
  };

  logInput(inputs, data.parameters);

  const headers = {
    "Content-Type": "application/json",
    "Authorization": "",
  };
  if(apiToken){
    headers.Authorization = `Bearer ${apiToken}`;
  }

  const res = await fetch(endpoint, {
    method: "POST",
    headers,
    body: JSON.stringify(data),
  });

  if(!res.ok){
    console.error("Error sending a request", res.status, res.statusText);
    setDefaultStatus();
    return null;
  }

  const generatedTextRaw = getGeneratedText(await res.json());

  let generatedText = generatedTextRaw;
  if(generatedText.slice(0, inputs.length) === inputs){
    generatedText = generatedText.slice(inputs.length);
  }
  const regexToClear = new RegExp([...stopTokens, ...tokensToClear].map(token => token.replace(/[.*+?^${}()|[\]\\]/g, '\\$&')).join('|'), 'g');
  generatedText = generatedText.replace(regexToClear, "");

  const resultEntry: ResultEntry = {
    new_prefix: generatedText,
    old_suffix: "",
    new_suffix: ""
  }

  const result: AutocompleteResult = {
    results: [resultEntry],
    old_prefix: "",
    user_message: [],
    is_locked: false,
  }

  setDefaultStatus();
  logOutput(generatedTextRaw);
  return result;
}

function getGeneratedText(json: any): string{
  return json?.generated_text ?? json?.[0].generated_text ?? "";
}

export type KnownLanguageType = keyof typeof languages;

export function getLanguageFileExtension(
  languageId: string
): string | undefined {
  return languages[languageId as KnownLanguageType];
}

export function getFileNameWithExtension(document: TextDocument): string {
  const { languageId, fileName } = document;
  if (!document.isUntitled) {
    return fileName;
  }
  const extension = getLanguageFileExtension(languageId);
  if (extension) {
    return fileName.concat(extension);
  }
  return fileName;
}

function clipMaxNewTokens(maxNewTokens: number): number {
  const MIN = 50;
  const MAX = 500;
  return Math.min(Math.max(maxNewTokens, MIN), MAX);
}

src/outputChannels.ts
import { WorkspaceConfiguration, window, workspace } from "vscode";
import * as fs from 'fs'
import { OUTPUT_CHANNEL_NAME } from "./globals/consts";

type Config = WorkspaceConfiguration & {
    logPath: string;
};

const outputChannel = window.createOutputChannel(OUTPUT_CHANNEL_NAME);

export function logInput(txt: string, parameters = {}){
    const config: Config = workspace.getConfiguration("HuggingFaceCode") as Config;
    const { logPath } = config;
    outputChannel.append(`INPUT to API: (with parameters ${JSON.stringify(parameters)}) \n`)
    outputChannel.append(txt);
    outputChannel.append("\n")
    if(logPath){
        fs.appendFileSync(logPath, JSON.stringify({time: Date.now(), type:"input", text: txt, parameters}) + "\n")
    }
}

export function logOutput(txt: string){
    const config: Config = workspace.getConfiguration("HuggingFaceCode") as Config;
    const { logPath } = config;
    outputChannel.append("OUTPUT from API:\n")
    outputChannel.append(txt);
    outputChannel.append("\n\n");

    if(logPath){
        fs.appendFileSync(logPath, JSON.stringify({time: Date.now(), type:"output", text: txt}) + "\n")
    }
}

export default outputChannel;

src/configTemplates.ts
const templateKeys = ["bigcode/starcoder", "codellama/CodeLlama-13b-hf", "Phind/Phind-CodeLlama-34B-v2", "WizardLM/WizardCoder-Python-34B-V1.0", "Custom"] as const;
export type TemplateKey = typeof templateKeys[number];

export interface TokenizerPathConfig {
    path: string;
}

export interface TokenizerRepoConfig {
    repository: string;
}

export interface TokenizerUrlConfig {
    url: string;
    to: string;
}

export interface Config {
    modelIdOrEndpoint: string;
    "fillInTheMiddle.enabled": boolean;
    "fillInTheMiddle.prefix": string;
    "fillInTheMiddle.middle": string;
    "fillInTheMiddle.suffix": string;
    temperature: number;
    contextWindow: number;
    tokensToClear: string[];
    tokenizer: TokenizerPathConfig | TokenizerRepoConfig | TokenizerUrlConfig | null;
}

const StarCoderConfig: Config = {
    modelIdOrEndpoint: "bigcode/starcoder",
    "fillInTheMiddle.enabled": true,
    "fillInTheMiddle.prefix": "<fim_prefix>",
    "fillInTheMiddle.middle": "<fim_middle>",
    "fillInTheMiddle.suffix": "<fim_suffix>",
    temperature: 0.2,
    contextWindow: 8192,
    tokensToClear: ["<|endoftext|>"],
    tokenizer: {
        repository: "bigcode/starcoder",
    }
}

const CodeLlama13BConfig: Config = {
    modelIdOrEndpoint: "codellama/CodeLlama-13b-hf",
    "fillInTheMiddle.enabled": true,
    "fillInTheMiddle.prefix": "<PRE> ",
    "fillInTheMiddle.middle": " <MID>",
    "fillInTheMiddle.suffix": " <SUF>",
    temperature: 0.2,
    contextWindow: 4096,
    tokensToClear: ["<EOT>"],
    tokenizer: {
        repository: "codellama/CodeLlama-13b-hf",
    }
}

const PhindCodeLlama34Bv2Config: Config = {
    ...CodeLlama13BConfig,
    modelIdOrEndpoint: "Phind/Phind-CodeLlama-34B-v2",
}

const WizardCoderPython34Bv1Config: Config = {
    ...CodeLlama13BConfig,
    modelIdOrEndpoint: "WizardLM/WizardCoder-Python-34B-V1.0",
    tokenizer: {
        repository: "WizardLM/WizardCoder-Python-34B-V1.0",
    }
}

export const templates: Partial<Record<TemplateKey, Config>> = {
    "bigcode/starcoder": StarCoderConfig,
    "codellama/CodeLlama-13b-hf": CodeLlama13BConfig,
    "Phind/Phind-CodeLlama-34B-v2": PhindCodeLlama34Bv2Config,
    "WizardLM/WizardCoder-Python-34B-V1.0": WizardCoderPython34Bv1Config,
}

src/inlineSuggestions/highlightStackAttributions.ts
import {
  Range,
  Selection,
  WorkspaceConfiguration,
  window,
  workspace,
  env,
  Uri
} from "vscode";

import fetch from "node-fetch";

export default async function highlightStackAttributions(): Promise<void> {
  const document = window.activeTextEditor?.document
  if (!document) return;

  type Config = WorkspaceConfiguration & {
    attributionWindowSize: number;
    attributionEndpoint: string;
  };
  const config: Config = workspace.getConfiguration("HuggingFaceCode") as Config;
  const { attributionWindowSize, attributionEndpoint } = config;

  // get cursor postion and offset
  const cursorPosition = window.activeTextEditor?.selection.active;
  if (!cursorPosition) return;
  const cursorOffset = document.offsetAt(cursorPosition);

  const start = Math.max(0, cursorOffset - attributionWindowSize);
  const end = Math.min(document.getText().length, cursorOffset + attributionWindowSize);

  // Select the start to end span
  if (!window.activeTextEditor) return;
  window.activeTextEditor.selection = new Selection(document.positionAt(start), document.positionAt(end));
  // new Range(document.positionAt(start), document.positionAt(end));
  

  const text = document.getText();
  const textAroundCursor = text.slice(start, end);

  const body = { document: textAroundCursor };

  // notify user request has started
  void window.showInformationMessage("Searching for nearby code in the stack...");
    
  const resp = await fetch(attributionEndpoint, {
    method: "POST",
    body: JSON.stringify(body),
    headers: { "Content-Type": "application/json" },
  });

  if(!resp.ok){
      return;
  }

  const json = await resp.json() as any as {spans: [number, number][]}      
  const {spans} = json

  if(spans.length === 0) {
    void window.showInformationMessage("No code found in the stack");
    return;
  }

  void window.showInformationMessage("Highlighted code was found in the stack.",
    "Go to stack search"
  ).then(clicked => {
    if (clicked) {
      // open stack search url in browser
      void env.openExternal(Uri.parse("https://huggingface.co/spaces/bigcode/search"));
    }
  });

  // combine overlapping spans
  const combinedSpans: [number,number][] = spans.reduce((acc, span) => {
    const [s, e] = span;
    if(acc.length === 0) return [[s, e]];
    const [lastStart, lastEnd] = acc[acc.length - 1];
    if(s <= lastEnd) {
      acc[acc.length - 1] = [lastStart, Math.max(lastEnd, e)];
    }else{
      acc.push([s, e]);
    }
    return acc;
  }, [] as [number, number][]);
  
  const decorations = combinedSpans.map(([startChar, endChar]) => ({range: new Range(document.positionAt(startChar + start), document.positionAt(endChar + start)), hoverMessage: "This code might be in the stack!"}))
  
  // console.log("Highlighting", decorations.map(d => [d.range.start, d.range.end]));

  const decorationType = window.createTextEditorDecorationType({
    color: 'red',
    textDecoration: 'underline',
    
  });

  window.activeTextEditor?.setDecorations(decorationType, decorations);

  setTimeout(() => {
    window.activeTextEditor?.setDecorations(decorationType, []);
  }, 5000);
}



src/extension.ts
import * as vscode from 'vscode';
import {
	LanguageClient,
	LanguageClientOptions,
	ServerOptions,
	TransportKind
} from 'vscode-languageclient/node';
import { TemplateKey, templates } from './configTemplates';
import { readFile } from 'fs';
import { homedir } from 'os';
import * as path from 'path';
import { fetch } from 'undici';

interface Completion {
	generated_text: string;
}

let client: LanguageClient;
let ctx: vscode.ExtensionContext;

export function activate(context: vscode.ExtensionContext) {
	ctx = context;
	handleConfigTemplateChange(ctx);
	const config = vscode.workspace.getConfiguration("llm");
	const binaryPath: string | null = config.get("lsp.binaryPath") as string | null;
	let command: string;
	if (binaryPath) {
		command = binaryPath;
	} else {
		const ext = process.platform === "win32" ? ".exe" : "";
		command = vscode.Uri.joinPath(context.extensionUri, "server", `llm-ls${ext}`).fsPath;
	}
	if (command.startsWith("~/")) {
		command = homedir() + command.slice("~".length);
	}
	console.log(`command: ${command}`);
	const serverOptions: ServerOptions = {
		run: {
			command, transport: TransportKind.stdio, options: {
				env: {
					"LLM_LOG_LEVEL": config.get("lsp.logLevel") as string,
				}
			}
		},
		debug: {
			command,
			transport: TransportKind.stdio,
			options: {
				env: {
					"LLM_LOG_LEVEL": config.get("lsp.logLevel") as string,
				}
			}
		}
	};

	const clientOptions: LanguageClientOptions = {
		documentSelector: [{ scheme: "file" }],
	};
	client = new LanguageClient(
		'llm',
		'LLM VS Code',
		serverOptions,
		clientOptions
	);

	client.start();

	const afterInsert = vscode.commands.registerCommand('llm.afterInsert', async (...args) => {
		// TODO: telemetry
	});
	ctx.subscriptions.push(afterInsert);

	const login = vscode.commands.registerCommand('llm.login', async (...args) => {
		const apiToken = await ctx.secrets.get('apiToken');
		if (apiToken !== undefined) {
			vscode.window.showInformationMessage('Llm: Already logged in');
			return;
		}
		const tokenPath = path.join(homedir(), path.sep, ".cache", path.sep, "huggingface", path.sep, "token");
		const token: string | undefined = await new Promise((res) => {
			readFile(tokenPath, (err, data) => {
				if (err) {
					res(undefined);
				}
				const content = data.toString();
				res(content.trim());
			});
		});
		if (token !== undefined) {
			await ctx.secrets.store('apiToken', token);
			vscode.window.showInformationMessage(`Llm: Logged in from cache: ~/.cache/huggingface/token ${tokenPath}`);
			return;
		}
		const input = await vscode.window.showInputBox({
			prompt: 'Please enter your API token (find yours at hf.co/settings/token):',
			placeHolder: 'Your token goes here ...'
		});
		if (input !== undefined) {
			await ctx.secrets.store('apiToken', input);
			vscode.window.showInformationMessage('Llm: Logged in succesfully');
		}
	});
	ctx.subscriptions.push(login);
	const logout = vscode.commands.registerCommand('llm.logout', async (...args) => {
		await ctx.secrets.delete('apiToken');
		vscode.window.showInformationMessage('Llm: Logged out');
	});
	ctx.subscriptions.push(logout);

	const attribution = vscode.commands.registerTextEditorCommand('llm.attribution', () => {
		void highlightStackAttributions();
	});
	ctx.subscriptions.push(attribution);
	const provider: vscode.InlineCompletionItemProvider = {
		async provideInlineCompletionItems(document, position, context, token) {
			console.log(context);
			if (position.line <= 0) {
				console.log("line <= 0");
				return;
			}

			const config = vscode.workspace.getConfiguration("llm");
			let params = {
				position,
				textDocument: client.code2ProtocolConverter.asTextDocumentIdentifier(document),
				model: config.get("modelIdOrEndpoint") as string,
				tokens_to_clear: config.get("tokensToClear") as string[],
				api_token: (await ctx.secrets.get('apiToken')) ?? "",
				request_params: {
					max_new_tokens: config.get("maxNewTokens") as number,
					temperature: config.get("temperature") as number,
					do_sample: true,
					top_p: 0.95,
				},
				fim: config.get("fillInTheMiddle") as number,
				context_window: config.get("contextWindow") as number,
				tls_skip_verify_insecure: config.get("tlsSkipVerifyInsecure") as boolean,
				ide: "vscode",
				tokenizer_config: config.get("tokenizer") as object | null,
			};
			try {
				const completions: Completion[] = await client.sendRequest("llm-ls/getCompletions", params, token);

				const items = [];
				for (const completion of completions as Completion[]) {
					items.push({
						insertText: completion.generated_text,
						range: new vscode.Range(position, position),
						command: {
							title: 'afterInsert',
							command: 'llm-vscode.afterInsert',
							arguments: [{ insertText: completion.generated_text }],
						}
					});
				}

				return {
					items,
				};
			} catch (e) {
				const err_msg = (e as Error).message;
				if (err_msg.includes("is currently loading")) {
					vscode.window.showWarningMessage(err_msg);
				} else if (err_msg !== "Canceled") {
					vscode.window.showErrorMessage(err_msg);
				}
			}

		},

	};
	vscode.languages.registerInlineCompletionItemProvider({ pattern: '**' }, provider);
}

export function deactivate() {
	if (!client) {
		return undefined;
	}
	return client.stop();
}

function handleConfigTemplateChange(context: vscode.ExtensionContext) {
	const listener = vscode.workspace.onDidChangeConfiguration(async event => {
		if (event.affectsConfiguration('llm.configTemplate')) {
			const config = vscode.workspace.getConfiguration("llm");
			const configKey = config.get("configTemplate") as TemplateKey;
			const template = templates[configKey];
			if (template) {
				const updatePromises = Object.entries(template).map(([key, val]) => config.update(key, val, vscode.ConfigurationTarget.Global));
				await Promise.all(updatePromises);
			}
		}
	});
	context.subscriptions.push(listener);
}

// TODO: refactor to select only highlighted code
export default async function highlightStackAttributions(): Promise<void> {
	const document = vscode.window.activeTextEditor?.document
	if (!document) return;

	const config = vscode.workspace.getConfiguration("llm");
	const attributionWindowSize = config.get("attributionWindowSize") as number;
	const attributionEndpoint = config.get("attributionEndpoint") as string;

	// get cursor postion and offset
	const cursorPosition = vscode.window.activeTextEditor?.selection.active;
	if (!cursorPosition) return;
	const cursorOffset = document.offsetAt(cursorPosition);

	const start = Math.max(0, cursorOffset - attributionWindowSize);
	const end = Math.min(document.getText().length, cursorOffset + attributionWindowSize);

	// Select the start to end span
	if (!vscode.window.activeTextEditor) return;
	vscode.window.activeTextEditor.selection = new vscode.Selection(document.positionAt(start), document.positionAt(end));
	// new Range(document.positionAt(start), document.positionAt(end));


	const text = document.getText();
	const textAroundCursor = text.slice(start, end);

	const body = { document: textAroundCursor };

	// notify user request has started
	void vscode.window.showInformationMessage("Searching for nearby code in the stack...");

	const resp = await fetch(attributionEndpoint, {
		method: "POST",
		body: JSON.stringify(body),
		headers: { "Content-Type": "application/json" },
	});

	if (!resp.ok) {
		return;
	}

	const json = await resp.json() as any as { spans: [number, number][] }
	const { spans } = json

	if (spans.length === 0) {
		void vscode.window.showInformationMessage("No code found in the stack");
		return;
	}

	void vscode.window.showInformationMessage("Highlighted code was found in the stack.",
		"Go to stack search"
	).then(clicked => {
		if (clicked) {
			// open stack search url in browser
			void vscode.env.openExternal(vscode.Uri.parse("https://huggingface.co/spaces/bigcode/search"));
		}
	});

	// combine overlapping spans
	const combinedSpans: [number, number][] = spans.reduce((acc, span) => {
		const [s, e] = span;
		if (acc.length === 0) return [[s, e]];
		const [lastStart, lastEnd] = acc[acc.length - 1];
		if (s <= lastEnd) {
			acc[acc.length - 1] = [lastStart, Math.max(lastEnd, e)];
		} else {
			acc.push([s, e]);
		}
		return acc;
	}, [] as [number, number][]);

	const decorations = combinedSpans.map(([startChar, endChar]) => ({ range: new vscode.Range(document.positionAt(startChar + start), document.positionAt(endChar + start)), hoverMessage: "This code might be in the stack!" }))

	// console.log("Highlighting", decorations.map(d => [d.range.start, d.range.end]));

	const decorationType = vscode.window.createTextEditorDecorationType({
		color: 'red',
		textDecoration: 'underline',

	});

	vscode.window.activeTextEditor?.setDecorations(decorationType, decorations);

	setTimeout(() => {
		vscode.window.activeTextEditor?.setDecorations(decorationType, []);
	}, 5000);
}